content/issue-2/applicative-parsing.md

```diff
@@ -0,0 +1,711 @@
+most of the functions would take a regular string reference as a filename - not every `String` is
+a valid filename and not every filename is a valid `String`. It's a good practice to keep
+`String` for strings and `PathBuf` for filenames. Type can also serve as a documentation for
+users.
+
+To be able to consume `PathBuf` from user's input parser needs a way to represent it in the first
+place. One approach would be to make `Magic` a generic datatype:
+
+```rust
+struct Magic<T>(T);
+impl<T> Magic<T> {
+    fn run(self) -> T {
+        self.0
+    }
+}
+```
+
+This allows to represent values and get them out but since handwavy magic creates variables of
+type `Magic<&'static str>` parser still needs a way to change the type to `Magic<PathBuf>` or
+any other type.
+
+For that Category Theory gives an abstraction called
+[`Functor`](https://en.wikipedia.org/wiki/Functor_(functional_programming)?useskin=vector).
+A functor allows to change a value inside of a generic context to some other value with the same
+or a different type but without changing the context. Sounds scary, but `Option::map` does more
+or less the same thing:
+
+```rust
+impl<T> Magic<T> {
+    fn map<R, F: Fn(T) -> R>(self, f: F) -> Magic<R> {
+        Magic(f(self.0))
+    }
+}
+
+#[test]
+fn sample() {
+    use std::path::PathBuf;
+    // --name Cargo.toml
+    let name = Magic("Cargo.toml"); // Magic<&str>
+    let _result = name.map(PathBuf::from).run(); // "Cargo.toml", PathBuf
+}
+```
+
+Every valid `Functor` implementation must satisfy a two laws - preserving identity morphism and
+preserving composition of morphisms. For as long as it only changes the values in a context
+without affecting the context itself - implementation should be valid.
+
+At this point parser can represent arguments of any type and provides a way to map between types
+
+## Handling failures
+
+Supporting mapping that can't fail is easy with just the `Functor` abstraction but it can't
+handle failures well - consider parsing numeric information:
+
+```
+$ app --answer 42
+$ app --answer Forty-two
+```
+
+A parser using `map`, `FromStr::from_str` and `unwrap` can parse the first line but panics
+with a bad error message for the second one. To handle parsing failures `Magic` needs to be
+able to represent them. The usual approach for dealing with failing computations in Rust is
+with help from the `Result` type.
+
+```rust
+struct Magic<T>(Result<T, String>);
+impl<T> Magic<T> {
+    fn run(self) -> Result<T, String> {
+        self.0
+    }
+
+    fn parse<R, F: Fn(T) -> Result<R, String>>(self, f: F) -> Magic<R> {
+        match self.0 {
+            Ok(x) => Magic(f(x)),
+            Err(err) => Magic(Err(err)),
+        }
+    }
+```
+
+Function `parse` applies a failing computation on a value inside `Magic` if one exists or keeps
+the error message untouched otherwise. Change in the representation requires to change `map` and
+`parse` can help with that too.
```

Comment:
```suggestion
The `parse` method applies a failing computation on a value inside `Magic` if one exists or keeps
the error message untouched otherwise. The change in the representation of `Magic` also 
requires to change `map`, but after that, fallible conversions are now possible:
```